### DISCUSSION TITLE
M2O and cascade delete

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#7659] [Archive]
Hi,

I have some issue understanding how I should set a cascade delete using the M2O relations.
I have a wharehouses collection that has a M2O relation on a locations collection.
```
wharehouse {
  id
  location : {
    id
  }
}
```
So I want to have a wharehouse that can have one location. I will also always query using this side.

I would like to configure that when I delete the wharehouse, the location will be deleted as well.

For now what I succeed is to have a M2O relation on the wharehouse to location, and a O2M relation on the locations to the wharehouse. On the wharehouse, I have configured the trigger "delete in cascade" of the location when the wharehouse is deleted. The advantage is I can select only one location for my wharehouse and I have also the delete in cascade.

Ideally I do not want to have and maintain the O2M relation as it is only usefull in my case to have my cascade delete.
Ideally as well, I do not want to configure for each M2O relation a delete in cascade using hooks.

A last point is that I would like to be able to create a wharehouse in one shot using the admin UI. 
1- start creation of the wharehouse
2- create the location using the "internal contextual panel" and link it to to the wharehouse that is on creation
3- finalize the creation of the wharehouse

I'm not sure also to understand why I cannot select a trigger on the M2O so when the entity is deleted the relation is as well deleted.

Thank you

**Comment by rijkvanzanten** [comment]
You can set the many to one up in one of two ways:

(A) `warehouses.location <-> locations.id`

of

(B) `warehouses.id <-> locations.warehouse`

(A) has the benefit that you can only have a single location per warehouse, but doesn't allow for deletion of the location when the warehouse is deleted.
(B) has the cascade delete capabilities, allowing you to delete the location when the warehouse is deleted, but also means that warehouses->locations is a _one to many_, allowing multiple locations to be hooked up to a single warehouse.

**Reply by gkielwasser** [reply]
Ok I see, thank you, 

As orginally I would have used O2O relations, I will stick with the solution (A). I do not want to handle locations array when I will query the wharehouses collection. Also it makes also more sense in my data model.

I will set a after delete hook that will handle cascade delete for the appropriate collections.

I'm curious in case we use M2O relations, is there any plan to support the delete of the orphan relations after an entity is deleted? Also do you think handling the cascade deletion with the hook seems appropriate?

Thanks !

**Reply by rijkvanzanten** [reply]
> As orginally I would have used O2O relations

How would you've setup the o2o relationship in SQL where both sides have a cascading delete?

From my understanding, the only way to support a pseudo-o2o (in sql) is by adding a regular many to one field, and then adding a unique constraint on the foreign key column to only allow a single row of that table to be linked, so

```
warehouses.location <-> locations.id
```

where you make `location` unique, so you can only assign a location to a warehouse once. However, that setup won't give you a cascade delete from either side afaik.

**Reply by gkielwasser** [reply]
I will keep M2O and the manual deletion of the childs, as it seems there is no perfect solution :) . 

It will be a bit repetitive task to code a hook to delete them each time, but I have to say for now I have not found something more smart.

Thanks for your feedback !

**Reply by rijkvanzanten** [reply]
> It will be a bit repetitive task to code a hook to delete them each time, but I have to say for now I have not found something more smart.

That sounds like a perfectly adequate solution to me! Your on-delete hook can simply clean out any other (semi-) related data in this fashion 🙂 

That being said, I'd love to hear more on how you think O2O relationships should ideally work, especially with the eye on the data model in the database directly

**Reply by gkielwasser** [reply]
Ok so finnally after exploring the schema I stopped my quest of doing something "automatic" for now.
I have made a hook that is handling the deletion of a list of "allowed to delete relations". So for some entity, I declare the allowed related collections to delete.

```
export const ALLOWED_CASCADE_DELETE = {
	// see further why next line is commented
	// craftsmen: ['wharehouses'],
	wharehouses: ['locations'],
	locations: ['addresses']
}

```

I have on Wharehouses collection a M2O to Craftsmen collection. I have configured the relational trigger so that when I delete the craftsmen the wharehouses are also deleted. This is working perfectly fine.
However, when the wharehouse is deleted by the relational trigger, the 'delete.before' hook is not called.

Shoudn't the hooks be called whatever the means the entity is deleted (by a relationnal trigger or not)?

Thanks

**Reply by rijkvanzanten** [reply]
> However, when the wharehouse is deleted by the relational trigger, the 'delete.before' hook is not called.
> Shoudn't the hooks be called whatever the means the entity is deleted (by a relationnal trigger or not)?

Good point! In this particular case, I'm not sure if that's something we can actually throw the hook for, as it's not directus that does that cascading delete, but the database itself. Therefore, it doesn't go through Directus' services and thus doesn't fire the hooks 🤔

**Reply by benhaynes** [reply]
Hmm, well that's an interesting issue... 🍿
